### PR TITLE
Add isolated network option

### PR DESCRIPTION
**Pull request by chenchun**

### CODE REVIEW

File: network.go [hunk 1]
```diff
@@ -479,6 +485,17 @@ func NetworkOptionPersist(persist bool) NetworkOption {
 	}
 }
 
+// NetworkOptionIsolatedNetwork returns an option setter to set if a network should be isolated
+func NetworkOptionIsolatedNetwork(isolated bool) NetworkOption {
```
Comment (aboch) [inline network.go]: Given the default is that libnetwork created networks are not isolated, can we change the method to not take any bool: `NetworkOptionIsolatedNetwork()`. 

Besides this, the changes LGTM.
Thanks

Comment (chenchun) [inline network.go]: Removed

File: api/api.go [hunk 2]
```diff
@@ -279,6 +280,11 @@ func procCreateNetwork(c libnetwork.NetworkController, vars map[string]string, b
 	processCreateDefaults(c, &create)
 
 	options := []libnetwork.NetworkOption{}
+	if len(create.NetworkOpts) > 0 {
+		if _, ok := create.NetworkOpts[netlabel.Isolated]; ok {
```
Comment (mavenugo) [inline api/api.go]: the keyword `isolated` doesn't seem to be the right one.
The networks are by default `isolated` from one another. The actual functionality that this PR provides is disabling external connectivity / default-gateway. I guess we should call it `defaultGateway` or `external` or something equivalent ? Given the fact that we might make the `DefaultGateway` configurable with its own plugin, we have to start thinking this in that direction.

File: network.go [hunk 3]
```diff
@@ -163,6 +163,7 @@ type network struct {
 	persist      bool
 	stopWatchCh  chan struct{}
 	drvOnce      *sync.Once
+	isolated     bool
```
Comment (mavenugo) [inline network.go]: Above comment is applicable here as well.

**Comment by mavenugo** [comment]
@chenchun my only comment atm is on the keyword `isolated`. It has different meaning in the context of existing functionality. Pls refer to https://github.com/docker/libnetwork/blob/master/drivers/bridge/bridge.go#L283 for the current meaning of what `isolated` mean in the current libnetwork code.


https://github.com/moby/libnetwork/blob/706968655f9281857ae24e01503422200dea208a/drivers/bridge/bridge.go#L281-L311

**Comment by chenchun** [comment]
Yes, with considering the meaning of network isolation to libnetwork, it may not be proper here.
ping @aboch who suggests to call it `isolated`.

**Comment by aboch** [comment]
@mavenugo @chenchun 
That's true that `isolated` may be confusing, but consider it's a term being used inside the bridge driver code only, not a general term in libnetwork package yet. When I proposed, I was thinking of a network isolated from both internal networks and external world.

If we want to change it, I'd go for `NoExternalConnectivity` or better `Disconnected`, since it describes the network's property. 

I'd avoid the default gateway term, it is confusing given there is already a default gateway service, which is being internally added (when needed) to make sure all networks are in pair, they can all the access the external world out of the box.

I like `Disconnected`, `netlabel.Disconnected, NetworkOptionDisconnected(),...` WDYT ?

**Comment by mavenugo** [comment]
@aboch omg `disconnected` ... NOOOO :)

**Comment by aboch** [comment]
Ok, I think I got a better one this time: `internal`.

But I'll let @mavenugo add more as there might be a plan to allow the user to choose which driver libnetwork will use to provide the external connectivity. If we go that route, then a simple flag may not be the right choice.

**Comment by mavenugo** [comment]
@aboch @chenchun lets go with `internal` as the name.

Lets worry about the external connectivity driver for a later stage when we actually have the backend logic fully ironed out.

@chenchun could you please update the PR. The rest of the changes LGTM

**Comment by chenchun** [comment]
Renamed isolated to internal.

**Comment by mavenugo** [comment]
Thanks @chenchun.
LGTM

**Comment by aboch** [comment]
LGTM

File: drivers/bridge/bridge.go [hunk 4]
```diff
@@ -490,6 +490,12 @@ func parseNetworkOptions(id string, option options.Generic) (*networkConfigurati
 		config.EnableIPv6 = val.(bool)
 	}
 
+	if val, ok := option[netlabel.Internal]; ok {
+		if internal, ok := val.(bool); ok && internal {
+			return nil, &driverapi.ErrNotImplemented{}
```
Comment (mavenugo) [inline drivers/bridge/bridge.go]: @chenchun are you planning on pushing the fix to remove this error and handling the bridge case appropriately ?
If am not wrong, this should be fairly simple change in bridge driver. I think this flag could just translate to disabling ip-masquarding, which we already have a flag to disable. Please let us know asap.

Comment (chenchun) [inline drivers/bridge/bridge.go]: Thanks for your suggestion, I'm taking a look.

**Pull request merged**